### PR TITLE
UCP/RNDV: added multiple rkeys on recv side

### PR DESCRIPTION
**Pull request by hoopoepg**
- added mem-pool datatype for rndv remote key
- access to request/rkey is wrapped by inline calls

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5225/ for details (Mellanox internal link).

**Comment by hoopoepg** [comment]
@brminich could you look at this?
here is no additional asserts

File: src/ucp/core/ucp_request.inl [hunk 1]
```diff
@@ -333,8 +333,57 @@ static UCS_F_ALWAYS_INLINE void ucp_request_send_tag_stat(ucp_request_t *req)
 }
 
 static UCS_F_ALWAYS_INLINE
-uct_rkey_bundle_t *ucp_tag_rndv_rkey(ucp_request_t *req)
+uct_rkey_bundle_t *ucp_tag_rndv_rkey_bundle(ucp_request_t *req, int idx)
 {
-    return &req->send.rndv_get.rkey_bundle;
+    return &req->send.rndv_get.rkey->rkey_bundle[idx];
```
[the remark this answers is not in the document]
Comment (hoopoepg) [reply src/ucp/core/ucp_request.inl]: added

File: src/ucp/core/ucp_request.inl [hunk 2]
```diff
@@ -333,8 +333,57 @@ static UCS_F_ALWAYS_INLINE void ucp_request_send_tag_stat(ucp_request_t *req)
...
+uct_rkey_t ucp_tag_rndv_rkey(ucp_request_t *req, int idx)
+{
+    return (req->send.rndv_get.rkey != NULL) ?
+           ucp_tag_rndv_rkey_bundle(req, idx)->rkey : UCT_INVALID_RKEY;
+}
+
+static UCS_F_ALWAYS_INLINE
+int ucp_tag_rndv_is_rkey_valid(ucp_request_t *req, int idx)
+{
+    return req->send.rndv_get.rkey != NULL &&
```
[the remark this answers is not in the document]
Comment (hoopoepg) [reply src/ucp/core/ucp_request.inl]: fixed

File: src/ucp/core/ucp_request.inl [hunk 3]
```diff
@@ -333,8 +333,57 @@ static UCS_F_ALWAYS_INLINE void ucp_request_send_tag_stat(ucp_request_t *req)
...
+           ucp_tag_rndv_rkey_bundle(req, idx)->rkey != UCT_INVALID_RKEY;
+}
+
+static UCS_F_ALWAYS_INLINE void
+ucp_request_rndv_get_create(ucp_request_t *req)
```
[the remark this answers is not in the document]
Comment (hoopoepg) [reply src/ucp/core/ucp_request.inl]: renamed

File: src/ucp/core/ucp_request.inl [hunk 4]
```diff
@@ -333,8 +333,57 @@ static UCS_F_ALWAYS_INLINE void ucp_request_send_tag_stat(ucp_request_t *req)
...
+{
+    int i;
+
+    ucs_trace_req("rendezvous-get create request %p", req);
+    req->send.rndv_get.rkey = ucs_mpool_get_inline(&(req->send.ep->worker)->rndv_get_mp);
```
[the remark this answers is not in the document]
Comment (hoopoepg) [reply src/ucp/core/ucp_request.inl]: removed

File: src/ucp/tag/rndv.c [hunk 5]
```diff
@@ -186,8 +187,8 @@ static void ucp_rndv_complete_rndv_get(ucp_request_t *rndv_req)
     UCS_PROFILE_REQUEST_EVENT(rreq, "complete_rndv_get", 0); // TODO
     ucp_request_complete_tag_recv(rreq, UCS_OK);
 
-    if (ucp_tag_rndv_rkey(rndv_req)->rkey != UCT_INVALID_RKEY) {
-        uct_rkey_release(ucp_tag_rndv_rkey(rndv_req));
+    if (ucp_tag_rndv_is_rkey_valid(rndv_req, 0)) {
```
[the remark this answers is not in the document]
Comment (hoopoepg) [reply src/ucp/tag/rndv.c]: updated

File: src/ucp/core/ucp_request.inl (hunk 1 above)
[the remark this answers is not in the document]
Comment (brminich) [reply src/ucp/core/ucp_request.inl]: why do you check `idx < UCP_MAX_LANES`, if rkey_bundle is limited by UCP_MAX_RNDV_LANES?

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5226/ for details (Mellanox internal link).

File: src/ucp/core/ucp_request.inl (hunk 1 above)
[the remark this answers is not in the document]
Comment (hoopoepg) [reply src/ucp/core/ucp_request.inl]: yep, fixed

**Comment by hoopoepg** [comment]
I will update commit message on squash

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5227/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
@hoopoepg need to resolve conflict

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5241/ for details (Mellanox internal link).

**Comment by hoopoepg** [comment]
@yosefe resolved

**Pull request merged**